### PR TITLE
fix: Reduce cookie modifications to stop infinite loops with CMP tools

### PR DESCRIPTION
**Pull request by benjackwhite**
## Changes

Hopefully should help https://github.com/PostHog/posthog/issues/13073

Some Consent Management Platforms (CMP) like Osano use a MutationObserver to watch for any changes to a cookie. In response they often change something about their UI. This causes infinite loops where Session Recordings records a mutation, updates a cookie, which triggers the CMP to modify the UI, which triggers a DOM mutation, which we record... etc.

This changes the "register" calls which are the main offender here to first check if any properties have actually changes before persisting to a cookie.

Tested with the playground html against Osano and it seemed to fix the issue.

## Checklist
- [ ] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -60,6 +60,7 @@
         "eslint-plugin-prettier": "^4.2.1",
         "eslint-plugin-react": "^7.30.1",
         "eslint-plugin-react-hooks": "^4.6.0",
+        "express": "^4.18.2",
```
Comment (benjackwhite) [inline package.json:63]: Dev dep for the playground server

File: src/posthog-persistence.ts [hunk 2]
```diff
@@ -122,6 +122,9 @@ export class PostHogPersistence {
     }
 
     save(): void {
+        // NOTE: Saving frequently causes issues with Recordings and CMP tools which
+        // observe cookie changes, and modify their UI, often causing infinite loops.
+        // As such callers of this should ideally check that the data has changed beforehand
```
Comment (pauldambra) [inline src/posthog-persistence.ts:127]:
```suggestion
/**
* NOTE: Saving frequently causes issues with Recordings and CMP tools which
* observe cookie changes, and modify their UI, often causing infinite loops.
* As such callers of this should ideally check that the data has changed beforehand
*/
    save(): void {
```

like this so people get intellisense?

File: playground/session-recordings/index.html [hunk 3]
```diff
@@ -1,67 +1,109 @@
 <html>
-<head>
-    <title>PostHog JS Snippet test</title>
-</head>
-<body>
-    <button data-cy-button>Some button</button>
+    <head>
+        <title>PostHog JS Snippet test</title>
+        <!-- Below can be enabled to test with a CMP like Osano that often causes infinite loop errors -->
```
Comment (pauldambra) [inline playground/session-recordings/index.html:4]:
```suggestion
        <!-- Below can be enabled to test with a Consent Management Platform (CMP) like Osano that often causes infinite loop errors -->
```

File: src/posthog-persistence.ts [hunk 4]
```diff
@@ -122,6 +122,9 @@ export class PostHogPersistence {
     }
 
     save(): void {
+        // NOTE: Saving frequently causes issues with Recordings and CMP tools which
```
Comment (pauldambra) [inline src/posthog-persistence.ts:125]:
```suggestion
        // NOTE: Saving frequently causes issues with Recordings and Consent Management Platform (CMP) tools which
```

**Review by pauldambra — APPROVED** [review]
couple of suggestions

**Pull request merged**